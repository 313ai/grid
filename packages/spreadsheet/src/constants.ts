import { Sheet } from "./Spreadsheet";
import { DATATYPE, CellDataFormatting } from "./types";
import { isNull } from "@rowsncolumns/grid";

export const COLUMN_HEADER_WIDTH = 46;
export const ROW_HEADER_HEIGHT = 24;
export const DEFAULT_ROW_HEIGHT = 20;
export const DEFAULT_COLUMN_WIDTH = 100;
export const DARK_MODE_COLOR = "rgb(26, 32, 44)";
export const DARK_MODE_COLOR_LIGHT = "#252E3E";

/**
 * Number to alphabet
 * @param i 
 */
export const number2Alpha = (i: number): string => {
  return (
    (i >= 26 ? number2Alpha(((i / 26) >> 0) - 1) : "") +
    "abcdefghijklmnopqrstuvwxyz"[i % 26 >> 0]
  ).toUpperCase();
};

/**
 * Create a new sheet
 * @param param0 
 */
export const createNewSheet = ({ count }: { count: number }): Sheet => ({
  id: uuid(),
  name: `Sheet${count}`,
  cells: {},
  activeCell: null,
  selections: [],
  scrollState: { scrollTop: 0, scrollLeft: 0 },
  columnSizes: {},
  rowSizes: {}
});

/**
 * UUID generator
 */
export const uuid = () =>
  "_" +
  Math.random()
    .toString(36)
    .substr(2, 9);

/**
 * Format a string
 * @param value 
 * @param datatype 
 * @param formatting 
 */
export const format = (
  value?: string,
  datatype?: DATATYPE,
  formatting?: CellDataFormatting
): string | undefined => {
  if (value === void 0 || isNull(value) || datatype !== DATATYPE.NUMBER) return value;
  if (!formatting) return value;
  let num = parseFloat(value);
  if (formatting.decimals) {
    value = num.toFixed(formatting.decimals);
  }
  if (formatting.percent) {
    value = (num * 100).toFixed(2);
  }
  if (formatting.currency) {
    value = `${formatting.currencySymbol || '$'}${num.toFixed(2)}`;
  }
  return "" + value;
};

/**
 * Detect datatype of a string
 * @param value 
 */
export const detectDataType = (value?: string): DATATYPE | undefined => {
  if (isNull(value)) return undefined
  if (!isNaN(Number(value))) return DATATYPE.NUMBER
  return undefined
}