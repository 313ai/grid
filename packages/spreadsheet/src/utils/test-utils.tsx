import React from "react";
import { render } from "@testing-library/react";
import { Stage, Layer } from "react-konva";
import { ThemeProvider, theme } from "@chakra-ui/core";
import Konva from "konva";

const KonvaWrapper = ({ children }) => {
  return (
    <Stage>
      <ThemeWrapper>
        <Layer>{children}</Layer>
      </ThemeWrapper>
    </Stage>
  );
};

const konvaRenderer = (ui) => render(ui, { wrapper: KonvaWrapper });
const domRenderer = (ui) => render(ui, { wrapper: ThemeWrapper });

export * from "@testing-library/react";
export { konvaRenderer, domRenderer };

export const ThemeWrapper = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
