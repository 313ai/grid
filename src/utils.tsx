import React from "react";
import { Line, Rect } from "react-konva";
import { ShapeConfig } from "konva/types/Shape";

interface BoxProps extends ShapeConfig {
  onCrosshairMouseDown?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseMove?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseUp?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Create a box with custom top/right/bottom/left colors and widths
 * @param param0
 */
export const createBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
}: BoxProps) => {
  const commonProps = {
    perfectDrawEnabled: false,
    shadowForStrokeEnabled: false,
    hitStrokeWidth: 0,
    listening: false,
    lineCap: "square",
  };
  const lines = [
    <Line
      points={[x, y, x + width, y]}
      stroke={strokeTopColor}
      strokeWidth={strokeTopWidth}
      key="top"
      {...commonProps}
    />,
    <Line
      points={[x + width, y, x + width, y + height]}
      stroke={strokeRightColor}
      strokeWidth={strokeRightWidth}
      key="right"
      {...commonProps}
    />,
    <Line
      points={[x + width, y + height, x, y + height]}
      stroke={strokeBottomColor}
      strokeWidth={strokeBottomWidth}
      key="bottom"
      {...commonProps}
    />,
    <Line
      points={[x, y + height, x, y]}
      stroke={strokeLeftColor}
      strokeWidth={strokeLeftWidth}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <Rect
          fill={fill}
          x={x}
          y={y}
          width={width}
          height={height}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};

export const FillHandle: React.FC<BoxProps> = ({
  x = 0,
  y = 0,
  stroke,
  onMouseDown,
}) => {
  if (x === 0 || y === 0) return null;
  const crossHairSize = 8;
  const crossHairBorderSize = 1;
  return (
    <div
      style={{
        position: "absolute",
        left: x - crossHairSize / 2,
        top: y - crossHairSize / 2,
        width: crossHairSize,
        height: crossHairSize,
        border: `${crossHairBorderSize}px white solid`,
        background: stroke,
        cursor: "crosshair",
      }}
      onMouseDown={onMouseDown}
    />
  );
};

export const createHTMLBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
  strokeStyle = "solid",
}: BoxProps) => {
  const commonProps = {};
  width = width - Math.floor(strokeWidth);
  y = y - Math.ceil(strokeWidth / 2);
  const lines = [
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: width,
        height: strokeTopWidth,
        borderWidth: 0,
        borderColor: strokeTopColor,
        borderTopWidth: strokeTopWidth,
        borderStyle: strokeStyle,
      }}
      key="top"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x + width,
        top: y,
        width: strokeRightWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeRightColor,
        borderRightWidth: strokeRightWidth,
        borderStyle: strokeStyle,
      }}
      key="right"
      {...commonProps}
    />,
    ,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y + height,
        width: width + strokeTopWidth,
        height: strokeBottomWidth,
        borderWidth: 0,
        borderColor: strokeBottomColor,
        borderBottomWidth: strokeBottomWidth,
        borderStyle: strokeStyle,
      }}
      key="bottom"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: strokeLeftWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeLeftColor,
        borderLeftWidth: strokeLeftWidth,
        borderStyle: strokeStyle,
      }}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <div
          style={{
            position: "absolute",
            top: y,
            left: x,
            height,
            width,
            backgroundColor: fill,
            userSelect: "none",
            pointerEvents: "none",
          }}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};
