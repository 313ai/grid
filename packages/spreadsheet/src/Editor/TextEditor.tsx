import React, {
  memo,
  forwardRef,
  useMemo,
  useState,
  useEffect,
  useImperativeHandle,
  useCallback
} from "react";
import {
  createEditor,
  Node,
  Transforms,
  NodeEntry,
  Range,
  Editor,
  Point
} from "slate";
import {
  Slate,
  Editable,
  withReact,
  ReactEditor,
  RenderLeafProps
} from "slate-react";
import { withHistory } from "slate-history";
import {
  Direction,
  KeyCodes,
  SelectionArea,
  castToString
} from "@rowsncolumns/grid";
import useShiftDown from "../hooks/useShiftDown";
import { useColorMode, useTheme, Box } from "@chakra-ui/core";
import { DARK_MODE_COLOR, FORMULA_FONT, FORMULA_FONT_SIZE } from "../constants";
import {
  normalizeTokens,
  tokenVocabulary,
  operatorTokenNames,
  getSelectionColorAtIndex
} from "./../formulas/helpers";
import { Token } from "fast-formula-parser/grammar/lexing";
import { current } from "immer";
import { css } from "@emotion/core";

export interface EditableProps {
  value?: React.ReactText;
  onChange: (value: string) => void;
  onSubmit: (value: string, direction?: Direction) => void;
  onCancel: (e: React.KeyboardEvent<HTMLDivElement>) => void;
  fontFamily: string;
  fontSize: number;
  scale: number;
  color: string;
  wrapping: any;
  horizontalAlign: any;
  underline?: boolean;
  onKeyDown?: (e: React.KeyboardEvent<HTMLDivElement>) => void;
  options?: string[];
  isFormulaMode?: boolean;
  autoFocus?: boolean;
  supportedFormulas?: string[];
}

export type RefAttribute = {
  ref?: React.Ref<EditableRef | null>;
};

export type EditableRef = {
  focus: () => void;
  updateSelection?: (sel: SelectionArea) => void;
};

export const cleanFunctionToken = (text: string) => {
  return text; //.replace(new RegExp(/\(|\)/, 'gi'), '')
};
const CHARS_BOUNDS = ["+", " ", "(", "=", ","];
// export const shouldShowFunctionSuggestion = (editor: Editor) => {
//   const { selection } = editor
//   if (selection && Range.isCollapsed(selection)) {
//     const [start] = Range.edges(selection)
//     const charBefore = Editor.before(editor, start, { unit: 'character' })
//     const range = Editor.range(editor, start, charBefore)
//     const str = Editor.string(editor, range)
//     return CHARS_BOUNDS.indexOf(str) !== -1
//   }
//   return false
// }

export const functionSuggestion = (tokens: Token[], editor: Editor) => {
  const { selection } = editor;
  if (selection && Range.isCollapsed(selection)) {
    const [start] = Range.edges(selection);
    const charBefore = Editor.before(editor, start, { unit: "character" });
    const range = Editor.range(editor, start, charBefore);
    const str = Editor.string(editor, range);
    const token = tokens.find(token => token.endColumn === start.offset);
    return token &&
      str !== "(" &&
      (token.tokenType.name === "Column" ||
        token.tokenType.name === "Name" ||
        token.tokenType.name === "Function")
      ? token
      : void 0;
  }
  return void 0;
};

export const isTokenACell = (token?: Token | null) => {
  return (
    token &&
    (token.tokenType.name === "Column" || token.tokenType.name === "Cell")
  );
};

export const getCurrentCursorOffset = (editor: Editor) => {
  const { selection } = editor;
  if (selection && Range.isCollapsed(selection)) {
    const [start] = Range.edges(selection);
    return start;
  }
  return void 0;
};

export const showCellSuggestions = (
  editor: Editor,
  tokens: Token[],
  prevToken?: Token,
  curToken?: Token,
  nextToken?: Token
) => {
  if (
    nextToken &&
    ["Number", "Cell", "Comma", "At", "OpenParen", "String"].includes(
      nextToken.tokenType.name
    )
  ) {
    return false;
  }
  const start = getCurrentCursorOffset(editor);
  if (
    prevToken &&
    start &&
    prevToken.tokenType.name === "Function" &&
    start.offset <= prevToken.endOffset
  ) {
    return false;
  }

  return prevToken && operatorTokenNames.includes(prevToken.tokenType.name);
};

export const isCurrentPositionACell = (editor: Editor, tokens: Token[]) => {
  const token = getCurrentToken(tokens, editor);
  return isTokenACell(token);
};

export const getCurrentToken = (tokens: Token[], editor: Editor) => {
  const { selection } = editor;
  if (selection && Range.isCollapsed(selection)) {
    const [start] = Range.edges(selection);
    return tokens.find(token => token.endColumn === start.offset);
  }
  return void 0;
};

export const getPreviousToken = (
  tokens: Token[],
  editor: Editor
): Token | undefined => {
  const start = getCurrentCursorOffset(editor);
  let i = 0;
  let token;
  if (start) {
    while (i < tokens.length) {
      token = tokens[i];
      if (token.startColumn > start.offset) {
        token = tokens[i - 1];
        break;
      }
      i++;
    }
    return token;
  }
};

export const getNextToken = (
  tokens: Token[],
  editor: Editor
): Token | undefined => {
  const start = getCurrentCursorOffset(editor);
  if (!start) return void 0;
  return tokens.find(token => {
    return start.offset === token.startOffset;
  });
};

export const createSlateChildren = (text: string) => {
  const tokens = normalizeTokens(text);
  let leafs = [];
  let prevToken = 0;
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i];
    leafs.push({
      text: (prevToken !== token.startOffset ? " " : "") + token.image,
      index: token.index,
      selection: !!token.sel
    });
    prevToken = token.endColumn;
  }
  return leafs;
};

const Leaf = ({ attributes, children, leaf }: RenderLeafProps) => {
  return (
    <span
      {...attributes}
      className={leaf.cursor ? "insert-range-indicator" : void 0}
      style={{
        color: leaf.color as string
      }}
    >
      {children}
    </span>
  );
};

const TextEditor: React.FC<EditableProps & RefAttribute> = memo(
  forwardRef((props, forwardedRef) => {
    const {
      value: initialValue,
      onChange,
      onSubmit,
      onCancel,
      fontFamily,
      fontSize,
      scale,
      color,
      wrapping,
      horizontalAlign,
      underline,
      onKeyDown,
      options,
      isFormulaMode,
      autoFocus,
      supportedFormulas = [],
      ...rest
    } = props;
    const [suggestionToken, setSuggestionToken] = useState<Token>();
    const [cursorToken, setCursorSuggestionToken] = useState<
      Point | undefined
    >();
    const [target, setTarget] = useState<Range | undefined>();
    const serialize = useCallback(
      (value?: React.ReactText): Node[] => {
        return (castToString(value) ?? "").split("\n").map((line: string) => {
          const children = [{ text: line }];
          return {
            children
          };
        });
      },
      [isFormulaMode]
    );
    const deserialize = useCallback((value: Node[]) => {
      return value
        .map(element => {
          // @ts-ignore
          return element.children.map(leaf => leaf.text).join("");
        })
        .join("\n");
    }, []);
    useImperativeHandle(
      forwardedRef,
      () => {
        return {
          focus: () => {
            ReactEditor.focus(editor);
          },
          updateSelection: (sel: SelectionArea | undefined) => {
            console.log("called");
          }
        };
      },
      []
    );
    const [value, setValue] = useState<Node[]>(() => serialize(initialValue));
    const editor = useMemo(() => withHistory(withReact(createEditor())), []);
    const moveToEnd = useCallback(() => {
      ReactEditor.focus(editor);
      document.execCommand("selectAll", false, undefined);
      // collapse selection to the end
      document.getSelection()?.collapseToEnd();
    }, []);
    useEffect(() => {
      if (autoFocus) {
        requestAnimationFrame(moveToEnd);
      }
    }, []);

    const defaultHighlightedIndex = isFormulaMode ? 0 : null;

    const {
      highlightedIndex,
      onKeyDown: onShiftDownKeyDown,
      setInputValue,
      menuRef,
      selectedItem,
      items,
      onMouseMove,
      onMouseDown,
      onClick,
      setSelectedItem
    } = useShiftDown({
      showAllIfEmpty: !isFormulaMode,
      defaultHighlightedIndex,
      initialInputValue: isFormulaMode ? "" : initialValue,
      initialIsOpen: true,
      initialSelectedItem: initialValue,
      options: isFormulaMode ? supportedFormulas : options,
      onChange: item => {
        if (isFormulaMode && item !== void 0) {
          const text = suggestionToken?.image;
          const start = getCurrentCursorOffset(editor);
          if (start) {
            Transforms.delete(editor, {
              at: start,
              distance: text?.length,
              reverse: true
            });

            Transforms.insertNodes(editor, [{ text: `${item as string}(` }]);
          }
          return;
        }
        onSubmit?.(item as string);
      }
    });
    const decorate = useCallback(
      (entry: NodeEntry<Node>) => {
        if (!isFormulaMode) {
          return [];
        }

        const ranges: any = [];
        const [node, path] = entry;
        const { text } = node;
        let offset = 0;
        if (isFormulaMode) {
          const tokens = normalizeTokens(text as string);
          if (cursorToken && tokens.length && cursorToken.path[0] === path[0]) {
            ranges.push({
              anchor: { path, offset: cursorToken.offset },
              focus: { path, offset: cursorToken.offset },
              cursor: true
            });
          }
          let prevToken: Token;
          tokens.forEach(token => {
            let add = token.startOffset - (prevToken?.endColumn ?? 0);
            ranges.push({
              anchor: { path, offset: offset + token.image.length + add },
              focus: { path, offset },
              selection: !!token.sel,
              color:
                token?.index !== void 0
                  ? getSelectionColorAtIndex(token.index)
                  : token.tokenType.name === "String"
                  ? "green"
                  : token.tokenType.name === "Number"
                  ? "#15c"
                  : void 0
            });

            prevToken = token;

            offset = offset + token.image.length + add;
          });
        }

        return ranges;
      },
      [isFormulaMode, cursorToken]
    );

    useEffect(() => {
      const normalizedValue = deserialize(value);
      if (!isFormulaMode) {
        setInputValue(normalizedValue);
      }
      onChange?.(normalizedValue);
    }, [value]);

    useEffect(() => {
      /* Update editor text if initialValue changes, when user enters text in formula bar */
      if (deserialize(value) !== initialValue) {
        setValue(serialize(initialValue));
      }
    }, [initialValue]);

    const theme = useTheme();
    const { colorMode } = useColorMode();
    const isLight = colorMode === "light";
    const borderColor = isLight
      ? theme.colors.gray[300]
      : theme.colors.gray[600];
    const inputColor = isLight ? DARK_MODE_COLOR : theme.colors.white;
    const dropdownBgColor = isLight
      ? theme.colors.white
      : theme.colors.gray[700];

    return (
      <>
        <div
          style={{
            fontFamily: isFormulaMode ? FORMULA_FONT : fontFamily,
            fontSize: (isFormulaMode ? FORMULA_FONT_SIZE : fontSize) * scale,
            width: "100%",
            height: "100%",
            padding: "0 1px",
            margin: 0,
            boxSizing: "border-box",
            borderWidth: 0,
            outline: "none",
            resize: "none",
            overflow: "hidden",
            verticalAlign: "top",
            background: "transparent",
            color: color,
            whiteSpace: "pre-wrap",
            textAlign: isFormulaMode ? "left" : horizontalAlign,
            lineHeight: "14px",
            textDecoration: underline ? "underline" : "none",
            cursor: "text",
            flex: 1
          }}
        >
          <Slate
            editor={editor}
            value={value}
            onChange={value => {
              setValue(value);
              if (isFormulaMode) {
                const start = getCurrentCursorOffset(editor);
                if (!start) {
                  return;
                }
                const from = Editor.before(editor, start, { unit: "block" });
                if (!from) {
                  return;
                }
                const range = Editor.range(editor, from, start);
                const line = Editor.string(editor, range);

                const tokens = normalizeTokens(line);
                const fnToken = functionSuggestion(tokens, editor);
                const curToken = getCurrentToken(tokens, editor);
                const prevToken = getPreviousToken(tokens, editor);
                const nextToken = getNextToken(tokens, editor);
                const showFnSuggestions = !!fnToken;
                const showCellSuggestion = showCellSuggestions(
                  editor,
                  tokens,
                  prevToken,
                  curToken,
                  nextToken
                );
                const isNewCell = !isCurrentPositionACell(editor, tokens);
                const cellTokenIndex = showCellSuggestion
                  ? curToken
                    ? isNewCell
                      ? curToken?.endColumn
                      : curToken?.startOffset
                    : getCurrentCursorOffset(editor)?.offset
                  : null;

                /* Update input value */
                if (showFnSuggestions) {
                  setSuggestionToken(fnToken);
                  setInputValue(fnToken?.image ?? "");
                } else {
                  setSuggestionToken(void 0);
                  setInputValue("");
                }

                setCursorSuggestionToken(
                  showCellSuggestion ? getCurrentCursorOffset(editor) : void 0
                );
              }
            }}
          >
            <Editable
              decorate={decorate}
              renderLeaf={props => <Leaf {...props} />}
              onKeyDown={(e: React.KeyboardEvent<HTMLDivElement>) => {
                const isShiftKey = e.nativeEvent.shiftKey;
                const isMetaKey =
                  e.nativeEvent.metaKey || e.nativeEvent.ctrlKey;
                const isFromSelection =
                  highlightedIndex !== null && items.length > 0;
                const text =
                  highlightedIndex !== null && items.length > 0
                    ? items[highlightedIndex]
                    : deserialize(value);

                // Enter key
                if (e.which === KeyCodes.Enter) {
                  if (isFormulaMode && isFromSelection) {
                    setInputValue("");
                    e.preventDefault();
                  } else {
                    /* Add a new line when Cmd/Ctrl key is pressed */
                    if (isMetaKey) {
                      editor.insertBreak();
                      return;
                    }
                    onSubmit?.(
                      text,
                      isShiftKey ? Direction.Up : Direction.Down
                    );

                    e.preventDefault();
                  }
                }

                if (e.which === KeyCodes.Escape) {
                  onCancel && onCancel(e);
                }

                if (e.which === KeyCodes.Tab) {
                  /* Trap focus inside the grid */
                  e.preventDefault();
                  onSubmit &&
                    onSubmit(
                      text,
                      isShiftKey ? Direction.Left : Direction.Right
                    );
                }
                /* Global handler */
                onKeyDown?.(e);

                /* Pass callback to shiftdown hook */
                onShiftDownKeyDown(e);
              }}
            />
          </Slate>
        </div>
        {items?.length ? (
          <Box
            ref={menuRef}
            width="auto"
            left="-2px"
            shadow="md"
            background={dropdownBgColor}
            pb={1}
            pt={1}
            position="absolute"
            top="100%"
            mt="2px"
            borderColor={borderColor}
            borderStyle="solid"
            borderWidth={1}
            minWidth="calc(100% + 4px)"
            maxHeight={400}
            overflow="auto"
          >
            {items.map((item, index) => {
              return (
                <Box
                  padding={1}
                  pl={1}
                  pr={1}
                  whiteSpace="nowrap"
                  color={inputColor}
                  cursor="pointer"
                  fontFamily={isFormulaMode ? FORMULA_FONT : fontFamily}
                  fontSize={`${(isFormulaMode ? FORMULA_FONT_SIZE : fontSize) *
                    scale}px`}
                  onMouseMove={() => onMouseMove?.(index)}
                  onMouseDown={onMouseDown}
                  onClick={() => onClick?.(item)}
                  key={item}
                  style={{
                    fontWeight: selectedItem === item ? "bold" : "normal",
                    backgroundColor:
                      highlightedIndex === index
                        ? isLight
                          ? theme.colors.gray[100]
                          : "rgba(255,255,255,0.06)"
                        : dropdownBgColor
                  }}
                >
                  {item}
                </Box>
              );
            })}
          </Box>
        ) : null}
      </>
    );
  })
);

export default TextEditor;
