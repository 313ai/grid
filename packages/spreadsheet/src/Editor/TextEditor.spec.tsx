import React from "react";
import { domRenderer, cleanup } from "./../utils/test-utils";
import TextEditor from "./TextEditor";
import { DEFAULT_FONT_FAMILY, DEFAULT_FONT_SIZE } from "../constants";

global.document.execCommand = jest.fn();

describe("TextEditor", () => {
  afterEach(cleanup);
  const position = { x: 0, y: 0, width: 200, height: 20 };
  const cell = { rowIndex: 1, columnIndex: 1 };
  const activeCell = { rowIndex: 1, columnIndex: 1 };
  const editorType = "text";
  const value = 'hello world'
  it("renders an editor", () => {    
    const renderGrid = () =>
      domRenderer(
        <TextEditor
          value={value}
          fontFamily={DEFAULT_FONT_FAMILY}
          fontSize={DEFAULT_FONT_SIZE}
          scale={1}
          color='black'
          wrapping
          horizontalAlign='left'
        />
      );
    expect(renderGrid).not.toThrow();
  });

  it("matches snapshot", () => {
    const { asFragment } = domRenderer(
      <TextEditor
          value={value}
          fontFamily={DEFAULT_FONT_FAMILY}
          fontSize={DEFAULT_FONT_SIZE}
          scale={1}
          color='black'
          wrapping
          horizontalAlign='left'
        />
    );
    expect(asFragment()).toMatchSnapshot();
  });
})