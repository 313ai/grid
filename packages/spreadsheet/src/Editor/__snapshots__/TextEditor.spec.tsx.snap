// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextEditor matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="font-family: Arial; font-size: 12px; font-weight: normal; width: 100%; height: 100%; padding: 0px 1px; margin: 0px; box-sizing: border-box; border-width: 0px; outline: none; resize: none; overflow: hidden; vertical-align: top; background: transparent; color: black; white-space: pre-wrap; text-align: left; line-height: normal; text-decoration: none; cursor: text; flex: 1;"
  >
    <div
      contenteditable="true"
      data-gramm="false"
      data-slate-editor="true"
      data-slate-node="value"
      role="textbox"
      style="outline: none; white-space: pre-wrap; word-wrap: break-word;"
    >
      <div
        data-slate-node="element"
        style="position: relative;"
      >
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          >
            <span
              data-slate-string="true"
            >
              hello world
            </span>
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
